google-batch workflow interruptions with "UNAVAILABLE: 502:Bad Gateway"

## Bug report 

When running workflows with google-batch, I sometimes get an `UNAVAILABLE: 502:Bad Gateway` error which causes a workflow to be interrupted regardless of my error strategy. This is relatively rare, seemingly random, and happens mostly for large runs (usually thousands of jobs). The issue might also not be linked to nextflow itself but the google-batch service.
Finally, it's probably not related, but for the specific run I shared below, I had many jobs that were preempted (~800 among over ~6k submitted jobs).

### Expected behavior and actual behavior

Expected: Normal execution of the workflow or more feedback about what the actual issue is.
Actual: Workflow interrupted

### Steps to reproduce the problem

Unable to reproduce, but happens fairly consistently for large scale runs (custom RNA-seq pipeline)

### Program output 
See attached [.nextflow.log
](https://github.com/nextflow-io/nextflow/files/13444648/bad_gateway_nextflow.log)

Relevant lines:
```
Nov-20 18:27:15.840 [Task submitter] DEBUG nextflow.processor.TaskProcessor - Handling unexpected condition for
  task: name=ROME_QUANT_HS:FASTQ_FILTER (PPMI-Phase1-IR1.51625.BL.0000430667.5104-SL-0473.L002); work-dir=gs://nxf-work/cedric/ROMEQuant/4b/98813e83210abb7fcba3c9c5261e40
  error [com.google.api.gax.rpc.UnavailableException]: io.grpc.StatusRuntimeException: UNAVAILABLE: 502:Bad Gateway
Nov-20 18:27:15.857 [Task submitter] ERROR nextflow.processor.TaskProcessor - Error executing process > 'ROME_QUANT_HS:FASTQ_FILTER (PPMI-Phase1-IR1.51625.BL.0000430667.5104-SL-0473.L002)'

Caused by:
  UNAVAILABLE: 502:Bad Gateway

com.google.api.gax.rpc.UnavailableException: io.grpc.StatusRuntimeException: UNAVAILABLE: 502:Bad Gateway
        at com.google.api.gax.rpc.ApiExceptionFactory.createException(ApiExceptionFactory.java:112)
        at com.google.api.gax.rpc.ApiExceptionFactory.createException(ApiExceptionFactory.java:41)
        at com.google.api.gax.grpc.GrpcApiExceptionFactory.create(GrpcApiExceptionFactory.java:86)
        at com.google.api.gax.grpc.GrpcApiExceptionFactory.create(GrpcApiExceptionFactory.java:66)
        at com.google.api.gax.grpc.GrpcExceptionCallable$ExceptionTransformingFuture.onFailure(GrpcExceptionCallable.java:97)
        at com.google.api.core.ApiFutures$1.onFailure(ApiFutures.java:67)
        at com.google.common.util.concurrent.Futures$CallbackListener.run(Futures.java:1132)
        at com.google.common.util.concurrent.DirectExecutor.execute(DirectExecutor.java:31)
        at com.google.common.util.concurrent.AbstractFuture.executeListener(AbstractFuture.java:1270)
        at com.google.common.util.concurrent.AbstractFuture.complete(AbstractFuture.java:1038)
        at com.google.common.util.concurrent.AbstractFuture.setException(AbstractFuture.java:808)
        at io.grpc.stub.ClientCalls$GrpcFuture.setException(ClientCalls.java:574)
        at io.grpc.stub.ClientCalls$UnaryStreamToFuture.onClose(ClientCalls.java:544)
        at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
        at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
        at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
        at com.google.api.gax.grpc.ChannelPool$ReleasingClientCall$1.onClose(ChannelPool.java:541)
        at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:576)
        at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:70)
        at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:757)
--
                at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:107)
                at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:323)
                at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1254)
                at groovy.lang.MetaClassImpl.invokeMethodClosure(MetaClassImpl.java:1042)
                at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1128)
                at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1030)
                at groovy.lang.Closure.call(Closure.java:427)
                at groovy.lang.Closure.call(Closure.java:406)
                at groovy.lang.Closure.run(Closure.java:498)
                ... 1 common frames omitted
Caused by: io.grpc.StatusRuntimeException: UNAVAILABLE: 502:Bad Gateway
        at io.grpc.Status.asRuntimeException(Status.java:539)
        ... 14 common frames omitted
Nov-20 18:27:15.866 [Task submitter] DEBUG nextflow.Session - Session aborted -- Cause: io.grpc.StatusRuntimeException: UNAVAILABLE: 502:Bad Gateway
```

### Environment 

* Nextflow version: `23.09.3-edge build 5888`
* Java version: `openjdk version "11.0.20.1" 2023-08-24`
* Operating system: `Linux nf-tower-main 5.15.0-1027-gcp #34~20.04.1-Ubuntu SMP Mon Jan 9 18:40:09 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux`
* Bash version: `zsh 5.8 (x86_64-ubuntu-linux-gnu)`

+1, this terminally interrupts workflows, and anecdotally seems to happen more on large workflows than small ones. I'm executing pipelines with `-resume` to pick up and (hopefully) finish jobs. I updated today to `23.11.0-edge` to resolve the Google Batch spot-instance pre-emption errors -- thanks for all your work on the Batch executor so far!

#### Steps to reproduce

I'm not able to reproduce this deterministically.

#### Stack trace
```
com.google.api.gax.rpc.UnavailableException: io.grpc.StatusRuntimeException: UNAVAILABLE: 502:Bad Gateway
        at com.google.api.gax.rpc.ApiExceptionFactory.createException(ApiExceptionFactory.java:112)
        at com.google.api.gax.rpc.ApiExceptionFactory.createException(ApiExceptionFactory.java:41)
        at com.google.api.gax.grpc.GrpcApiExceptionFactory.create(GrpcApiExceptionFactory.java:86)
        at com.google.api.gax.grpc.GrpcApiExceptionFactory.create(GrpcApiExceptionFactory.java:66)
        at com.google.api.gax.grpc.GrpcExceptionCallable$ExceptionTransformingFuture.onFailure(GrpcExceptionCallable.java:97)
        at com.google.api.core.ApiFutures$1.onFailure(ApiFutures.java:84)
        at com.google.common.util.concurrent.Futures$CallbackListener.run(Futures.java:1127)
        at com.google.common.util.concurrent.DirectExecutor.execute(DirectExecutor.java:31)
        at com.google.common.util.concurrent.AbstractFuture.executeListener(AbstractFuture.java:1286)
        at com.google.common.util.concurrent.AbstractFuture.complete(AbstractFuture.java:1055)
        at com.google.common.util.concurrent.AbstractFuture.setException(AbstractFuture.java:807)
        at io.grpc.stub.ClientCalls$GrpcFuture.setException(ClientCalls.java:568)
        at io.grpc.stub.ClientCalls$UnaryStreamToFuture.onClose(ClientCalls.java:538)
        at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
        at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
        at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
        at com.google.api.gax.grpc.ChannelPool$ReleasingClientCall$1.onClose(ChannelPool.java:570)
        at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:567)
        at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:71)
        at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:735)
        at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:716)
        at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
        at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:133)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:829)
        Suppressed: com.google.api.gax.rpc.AsyncTaskException: Asynchronous task failed
                at com.google.api.gax.rpc.ApiExceptions.callAndTranslateApiException(ApiExceptions.java:57)
                at com.google.api.gax.rpc.UnaryCallable.call(UnaryCallable.java:112)
                at com.google.cloud.batch.v1.BatchServiceClient.createJob(BatchServiceClient.java:316)
                at com.google.cloud.batch.v1.BatchServiceClient.createJob(BatchServiceClient.java:248)
                at nextflow.cloud.google.batch.client.BatchClient.submitJob(BatchClient.groovy:85)
                at nextflow.cloud.google.batch.GoogleBatchTaskHandler.submit(GoogleBatchTaskHandler.groovy:136)
                at nextflow.processor.TaskPollingMonitor.submit(TaskPollingMonitor.groovy:196)
                at nextflow.processor.TaskPollingMonitor.submitPendingTasks(TaskPollingMonitor.groovy:565)
                at nextflow.processor.TaskPollingMonitor.submitLoop(TaskPollingMonitor.groovy:390)
                at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
                at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
                at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
                at java.base/java.lang.reflect.Method.invoke(Method.java:566)
                at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:107)
                at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:323)
                at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1254)
                at groovy.lang.MetaClassImpl.invokeMethodClosure(MetaClassImpl.java:1042)
                at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1128)
                at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1030)
                at groovy.lang.Closure.call(Closure.java:427)
                at groovy.lang.Closure.call(Closure.java:406)
                at groovy.lang.Closure.run(Closure.java:498)
                ... 1 common frames omitted
Caused by: io.grpc.StatusRuntimeException: UNAVAILABLE: 502:Bad Gateway
        at io.grpc.Status.asRuntimeException(Status.java:537)
        ... 14 common frames omitted
Dec-05 21:11:39.901 [Task submitter] DEBUG nextflow.Session - Session aborted -- Cause: io.grpc.StatusRuntimeException: UNAVAILABLE: 502:Bad Gateway
```

#### Environment
```
NextFlow version 23.11.0-edge build 5891
openjdk version "11.0.19" 2023-04-18
OS: Ubuntu 18.04.6 LTS
shell: GNU bash, version 4.4.20(1)-release (x86_64-pc-linux-gnu)
```

#### NF config:
```
google {
    project = 'xxx'
    region = 'us-central1,us-east1,us-east4,us-west1,us-west2'
    batch.preemptible = true
    batch.bootDiskSize = '20 GB'
    batch.debug = true
}

executor {
    pollInterval = '5 sec'
    dumpInterval = '10 sec'
    submitRateLimit = '10 sec'
}
```

Some jobs for the same sample that succeeded: nf-6ee35e49-1701807374390, nf-6b52752f-1701807374422, nf-6a013421-1701806533653. As above, the log does not contain a `nf-` ID for the failing job. It would be great if this were caught and handled as an individual task-level failure, so it could be governed by retry policy.

--- a/modules/nextflow/src/main/groovy/nextflow/executor/BashWrapperBuilder.groovy
+++ b/modules/nextflow/src/main/groovy/nextflow/executor/BashWrapperBuilder.groovy
@@ -449,7 +449,7 @@ class BashWrapperBuilder {
                 }
                 return path
             }
-            catch (FileSystemException | SocketException | RuntimeException e) {
+            catch (FileSystemException | IOException | RuntimeException | SocketException e) {
                 final isLocalFS = path.getFileSystem()==FileSystems.default
                 // the retry logic is needed for non-local file system such as S3.
                 // when the file is local fail without retrying

--- a/modules/nextflow/src/main/groovy/nextflow/executor/BashWrapperBuilder.groovy
+++ b/modules/nextflow/src/main/groovy/nextflow/executor/BashWrapperBuilder.groovy
@@ -449,7 +449,9 @@ class BashWrapperBuilder {
                 }
                 return path
             }
-            catch (FileSystemException | IOException | RuntimeException | SocketException e) {
+            catch (Exception e) {
+                if( !isRetryable0(e) )
+                    throw e
                 final isLocalFS = path.getFileSystem()==FileSystems.default
                 // the retry logic is needed for non-local file system such as S3.
                 // when the file is local fail without retrying
@@ -461,6 +463,18 @@ class BashWrapperBuilder {
                 Thread.sleep(delay)
             }
         }
+    }
+
+    static protected boolean isRetryable0(Exception e) {
+        if( e instanceof FileSystemException )
+            return true
+        if( e instanceof SocketException )
+            return true
+        if( e instanceof RuntimeException )
+            return true
+        if( e.class.getSimpleName() == 'HttpResponseException' )
+            return true
+        return false
     }
 
     protected String getTaskMetadata() {

--- a/modules/nextflow/src/test/groovy/nextflow/executor/BashWrapperBuilderTest.groovy
+++ b/modules/nextflow/src/test/groovy/nextflow/executor/BashWrapperBuilderTest.groovy
@@ -16,6 +16,7 @@
 
 package nextflow.executor
 
+import java.nio.file.FileSystemException
 import java.nio.file.Files
 import java.nio.file.Path
 import java.nio.file.Paths
@@ -1358,5 +1359,18 @@ class BashWrapperBuilderTest extends Specification {
         binding.launch_cmd == 'podman run -i -e "NXF_TASK_WORKDIR" -v /work/dir:/work/dir -v "$NXF_TASK_WORKDIR":"$NXF_TASK_WORKDIR" -w "$NXF_TASK_WORKDIR" --name $NXF_BOXID busybox /bin/bash -ue /work/dir/.command.sh'
         binding.cleanup_cmd == 'rm -rf $NXF_SCRATCH || true\npodman rm $NXF_BOXID &>/dev/null || true\n'
         binding.kill_cmd == 'podman stop $NXF_BOXID'
+    }
+
+    @Unroll
+    def 'should check retryable errors' () {
+        expect:
+        BashWrapperBuilder.isRetryable0(ERROR) == EXPECTED
+        where:
+        ERROR                           | EXPECTED
+        new RuntimeException()          | true
+        new SocketException()           | true
+        new FileSystemException('foo')  | true
+        new IOException()               | false
+        new Exception()                 | false
     }
 }

--- a/plugins/nf-google/src/test/nextflow/executor/BashWrapperBuilderWithGoogleTest.groovy
+++ b/plugins/nf-google/src/test/nextflow/executor/BashWrapperBuilderWithGoogleTest.groovy
@@ -1,0 +1,38 @@
+/*
+ * Copyright 2013-2024, Seqera Labs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package nextflow.executor
+
+
+import com.google.api.client.http.HttpResponseException
+import spock.lang.Specification
+/**
+ *
+ * @author Paolo Di Tommaso <paolo.ditommaso@gmail.com>
+ */
+class BashWrapperBuilderWithGoogleTest extends Specification {
+
+    def 'should check retryable errors' () {
+        expect:
+        BashWrapperBuilder.isRetryable0(ERROR) == EXPECTED
+        where:
+        ERROR                                                                   | EXPECTED
+        new HttpResponseException(GroovyMock(HttpResponseException.Builder))    | true
+        new Exception()                                                         | false
+    }
+
+}

